FlowCrypt/Functionality/Services/Backup Services/BackupService.swift

```diff
@@ -45,53 +45,56 @@ extension BackupService: BackupServiceType {
         }
     }
 
-    func backupToInbox(keys: [KeyDetails], for userId: UserId) -> Promise<Void> {
-        Promise { [weak self] resolve, reject -> Void in
-            guard let self = self else { throw AppErr.nilSelf }
-
-            let isFullyEncryptedKeys = keys.map(\.isFullyDecrypted).contains(false)
-
-            guard isFullyEncryptedKeys else {
-                throw BackupServiceError.keyIsNotFullyEncrypted
-            }
+    func backupToInbox(keys: [KeyDetails], for userId: UserId) -> Future<Void, Error> {
+        Future { [weak self] promise in
+            DispatchQueue.global().async {
```

Comment:
This is a responsibility for subscriber to decide on which queue to run this Future.
You can subscribe on a dedicated queue
`self.backupService.backupToInbox(keys: [encryptedPrv.key], for: self.user).subscribe(on: DispatchQueue.global()).wait()`

Reply:
Subscribe will work only for downstream, not upstream. You could check this with playground and following example:
```
func f1(_ value: Int) -> Future<Int, Error> {
    Future { promise in
        print("\(Thread.current): f1() value = \(value)")
        promise(.success(value + 1))
    }
}

func f2(_ value: Int) -> Future<Int, Error> {
    Future { promise in
        print("\(Thread.current): f2() value = \(value)")
        promise(.success(value + 1))
    }
}

let a2 = f1(0)
    .subscribe(on: DispatchQueue.global())
    .flatMap(f2)
    .receive(on: DispatchQueue.main)
    .sink { _ in
        print("\(Thread.current): completed")
    } receiveValue: { value in
        print("\(Thread.current): sink() value = \(value)")
    }
```

Output will be:
```
<_NSMainThread: 0x60000045c540>{number = 1, name = main}: f1() value = 0
<NSThread: 0x600000459d40>{number = 5, name = (null)}: f2() value = 1
<_NSMainThread: 0x60000045c540>{number = 1, name = main}: sink() value = 2
<_NSMainThread: 0x60000045c540>{number = 1, name = main}: completed
```

Reply:
So this is the reason why `Just` was used in my original code.

Reply:
What do you mean by `downstream, not upstream`? 
A Future will execute immediately when you create it.
If you need to postpone subscription use `Deffered` for this
https://developer.apple.com/documentation/combine/deferred

```
func f1(_ value: Int) -> AnyPublisher<Int, Error> {
    Deferred {
        Future { promise in
            sleep(1)
            let new = value + 1
            print("isMainThread \(Thread.isMainThread): f1() value = \(new)")
            promise(.success(new))
        }
    }
    .eraseToAnyPublisher()
}

func f2(_ value: Int) -> Future<Int, Error> {
    Future { promise in
        sleep(1)
        let new = value + 1
        print("isMainThread \(Thread.isMainThread): f2() value = \(value)")
        promise(.success(new))
    }
}

let a2 = f1(0)
    .subscribe(on: DispatchQueue.global())
    .flatMap(f2)
    .receive(on: DispatchQueue.main)
    .sink { _ in
        print("isMainThread \(Thread.isMainThread): completed")
    } receiveValue: { value in
        print("isMainThread \(Thread.isMainThread): sink() value = \(value)")
    }
```

Reply:
So you want to add `Deffered` in any method? Looks like a hack.

Reply:
@sosnovsky do you have any thoughts or preference regarding the two approaches above?

Reply:
My own design goal here is that the resulting code structure in Swift reads similar to async/await syntax in TypeScript and Kotlin, because these are the tree languages we use at our company. Doing it similarly adds a bit of consistency across codebases, and makes it easier for team members to jump across codebases, and also makes PR reviews easier.

So that would be something like (TS example)

```ts
async function bigAsyncMethod(input): Promise<X> {
  const thing = await something(input);
  if (thing > 10) {
    return await thingWasbig()
  }
  return await thingWasSmall()
}
```

which is then called like

```ts
bigAsyncMethod('hello').then(success => {
  // happy path
}).catch(error => {
  // error path
})
```

With the PR as currently written, I think we've achieved that goal. What's left to do is to settle on a good-practice code pattern for these situations in Swift, as Ivan and Anton are discussing.

Reply:
As it looks now, I may slightly prefer the `Deferred` approach as it seems to make usage of the methods a bit more flexible. But I'm not in a position to know if it's a good idea or not, since Swift is not my area.

Reply:
Yes, basically there is a responsibility for the subscriber of the Future to decide on which thread this publisher should be executed. 
Combine in general is an SDK to handle async operations, so there is a misusage of it in case you manually switch to some thread inside implementation. This breaks SOLID principles as well
You can make some convenience extension for Publisher which wraps Defferd and Promise and use this instead

Reply:
>  You can make some convenience extension for Publisher which wraps Defferd and Promise and use this instead

Do you mean it would wrap `Deferred` and `Future` ? Roughly what could it look like?

Reply:
> My own design goal here is that the resulting code structure in Swift reads similar to async/await syntax in TypeScript and Kotlin, because these are the tree languages we use at our company.

True async/await is available only with iOS 15 and coroutines. My `wait` method blocks current thread while real `await` should not. I'm not sure it is possible to get the same behaviour as TypeScript/Kotlin has.

Reply:
> True async/await is available only with iOS 15 and coroutines. My wait method blocks current thread while real await should not. I'm not sure it is possible to get the same behaviour as TypeScript/Kotlin has.

I understand that, but you had to add the `.wait()` because you were mixing Promise and Future, right? Once we are all Future, you shouldn't have to do that?

Reply:
```
extension Future {
    static func deferred(_ attempt: @escaping ((Result<Output, Failure>) -> Void) -> Void) -> AnyPublisher<Output, Failure> {
        return Deferred {
            Future.init(attempt)
        }
        .eraseToAnyPublisher()
    }
}
```

and usage
```
func f1(_ value: Int) -> AnyPublisher<Int, Error> {
    Future.deferred { promise in
        sleep(1)
        let new = value + 1
        print("isMainThread! \(Thread.isMainThread): f1() value = \(new)")
        promise(.success(new))
    }
}
```

Or any other solution, like extending AnyPublisher, creating your own Publisher - DefferedFuture. Or so...

Reply:
> Got it. Did you have to add the `.wait()` because you were mixing Promise and Future? Once we are all Future, you shouldn't have to do that?

Since you want to use `await` pattern and don't like to chain functions with `flatMap` (like I did in original code), we have to wait for future result somehow. So we have to use `wait`.

Reply:
@tomholub I usually prefer method chaining, as it was in original Ivan's code, it's quite common pattern for Swift. But after reading all comments here I think `wait` solution should be better for this case - to make code similar between different platforms. Also in the future it'll be easier to convert this code to native iOS `async/await`

Reply:
Thanks Roma

@Kharchevskyi is it not possible to put Deferred a the call point rather than into the method? It feels weird to have the first called method be "special" as you did it here: https://github.com/FlowCrypt/flowcrypt-ios/pull/590#discussion_r722482417 which would mean we would be putting Deferred inside all async method which seems also weird.

Also I want to avoid method chaining if possible.

So can it be something like this?

```swift
func f1(_ value: Int) -> Future<Int, Error> {
        Future { promise in
            sleep(1)
            let new = value + 1
            print("isMainThread \(Thread.isMainThread): f1() value = \(new)")
            promise(.success(new))
        }
}

func f2(_ value: Int) -> Future<Int, Error> {
    Future { promise in
        sleep(1)
        let new = value + 1
        print("isMainThread \(Thread.isMainThread): f2() value = \(value)")
        promise(.success(new))
    }
}
```

```Swift
// pseudo code
Future.deferred {
  let a = f1(0)
  let b = f2(a + 3)
}
.subscribe(on: DispatchQueue.global())
.receive(on: DispatchQueue.main)
.sink { _ in
  print("isMainThread \(Thread.isMainThread): completed")
} receiveValue: { value in
  print("isMainThread \(Thread.isMainThread): sink() value = \(value)")
}
```

Is there really no such thing before real async/await of iOS15? Do I have to wait every time and block the thread like this?

```swift
// pseudo code
Future.deferred {
  let a = f1(0).wait()
  let b = f2(a + 3).wait()
}
.subscribe(on: DispatchQueue.global())
.receive(on: DispatchQueue.main)
.sink { _ in
  print("isMainThread \(Thread.isMainThread): completed")
} receiveValue: { value in
  print("isMainThread \(Thread.isMainThread): sink() value = \(value)")
}
```